Make JitDump label output multi-thread safe

kunalspathak opened the pull request:
- Interlock the static variable when incrementing
- Fixed 2 broken links inside botr's `ryujit-overview.md` doc.

Fixes: #10984

kunalspathak committed:
Make JitDump label output multi-thread safe

`s_compMethodsCount` is a static variable and incrementing it when
two JITs are running at the same time can result in inconsistent
labels. This PR fixes it by incrementing it using
`InterlockedIncrement()`.

Fix #10984

kunalspathak committed:
Fix two links inside botr

Fixed the two broken links pointing to `viewing-jit-dumps.md`
inside botr.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=32578) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=32578) before we can accept your contribution.<br/><br/>:x: kunalspathak [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=32578)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=32578) it.</sub>

kunalspathak commented:
cc @BruceForstall

kunalspathak commented:
Looks like `<winnt.h>` doesn't get included. I will take a look tomorrow.

BruceForstall commented:
The issue isn't really the lack of Interlocked; it's that lots of places use `s_compMethodsCount`, but that could be updated by concurrent JIT invocations. We need to use an instance variable -- probably on our single-instance "Compiler" class, but potentially also a TLS variable.

As for winnt.h, we don't use that because we must be portable to Linux/Mac. We wrap everything with either utilcode.h/winwrap.h, or for Linux we use src\coreclr\src\pal\inc\pal.h.

kunalspathak committed:
Make an instance variable `compMethodID` of `Compiler` class

Added an instance variable `compMethodID` in `Compiler` class so it
can be used to print the method id.

kunalspathak committed:
Fix broken link inside debugging.md

Fix another broken link inside `debugging.md`.

kunalspathak committed:
Changed the type of s_compMethodID

Changed the type of s_compMethodID to `LONG` and made it private.

kunalspathak committed:
Changes in arm emitter and codegenarm

Changes in  emitter and codegenarm files of arm arch.

kunalspathak committed:
Ran jit-format tool

Run and fixed the jit-format tool.

BruceForstall commented on src/coreclr/src/jit/compiler.h:
```diff
@@ -8979,10 +8979,16 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
-    unsigned        compStatementID;
-    unsigned        compBasicBlockID;
+private:
+    static LONG s_compMethodID; // to produce unique label names
```
Shouldn't this still be `s_compMethodsCount`? After all, it is still a static global count of the number of methods the JIT has compiled. We just happen to often initialize the "method ID" using this.

kunalspathak replied:
You are right, it is still a static global method count that are JITted but it was (and is) used in producing a unique label names. In fact, I named the instance variable as `compMethodID` and it felt little odd to have name inconsistency during assignment. But I don't have any preference one way or other. I will go ahead and revert the name of this variable.

kunalspathak replied:
done.

BruceForstall reviewed (commented):
One question

BruceForstall commented:
cc @dotnet/jit-contrib

kunalspathak committed:
Rename static global variable

Revert the name of static global variable to `s_compMethodsCount`.

erozenfeld reviewed (approved):
LGTM

sandreenko commented on src/coreclr/src/jit/compiler.cpp:
```diff
@@ -1299,7 +1299,7 @@ size_t genFlowNodeCnt;
 #ifdef DEBUG
 /* static */
-unsigned Compiler::s_compMethodsCount = 0; // to produce unique label names
+LONG Compiler::s_compMethodsCount = 0; // to produce unique label names
```
have not seen many `LONG` in the jit, I am not sure in which case we use them.

BruceForstall replied:
@kunalspathak Presumably it chosen because it matches the signature of InterlockedIncrement

kunalspathak replied:
That's correct.

sandreenko reviewed (approved):
LGTM

benaadams commented:
I think there is a missed method (perhaps from overlapping merges)

https://github.com/dotnet/runtime/blob/4f84429fb44801b841c052505935a3cd14da18e2/src/coreclr/src/jit/codegencommon.cpp#L11307-L11312

Not sure runtime currently compiles?
